File: scala/unwrittenfun/minecraft/unwrittenblocks/recipes/InfuserRecipes.scala
package unwrittenfun.minecraft.unwrittenblocks.recipes

import scala.collection.mutable
import net.minecraft.item.ItemStack
import net.minecraft.item.crafting.FurnaceRecipes

/**
 * Mod: UnwrittenBlocks
 * Author: UnwrittenFun
 * License: Minecraft Mod Public License (Version 1.0.1)
 */
object InfuserRecipes {
  private val infusingList = new mutable.HashMap[List[Int], ItemStack]()
  private val ticksList = new mutable.HashMap[List[Int], Int]()
  private val ignoreMetaList = new mutable.HashMap[Int, Boolean]()

  // Set meta to -1 to ignore meta
  def addRecipe(id: Int, meta: Int, result: ItemStack, ticks: Int) {
    infusingList.put(List(id, meta), result)
    ticksList.put(List(id, meta), ticks)
    ignoreMetaList.put(id, meta == -1)
  }

  def getTickTime(id: Int, meta: Int): Int = {
    ticksList.get(List(id, getMeta(id, meta))).get
  }

  def getInfusingResult(id: Int, meta: Int): ItemStack = {
    if (infusingList.contains(List(id, getMeta(id, meta)))) {
      infusingList.get(List(id, getMeta(id, meta))).get
    } else {
      null
    }
  }

  // Return meta or -1 if should ignore meta
  private def getMeta(id: Int, meta: Int): Int = {
    if (!ignoreMetaList.contains(id) || ignoreMetaList.get(id).get) -1 else meta
  }
}

File: src/main/java/unwrittenfun/minecraft/unwrittenblocks/common/recipes/InfuserRecipes.java
package unwrittenfun.minecraft.unwrittenblocks.common.recipes;

import net.minecraft.item.ItemStack;

import java.util.HashMap;
import java.util.Map;

/**
 * Project: UnwrittenBlocks
 * Author: UnwrittenFun
 * Created: 05/11/2014.
 */
public class InfuserRecipes {
  public static InfuserRecipes instance = new InfuserRecipes();

  public HashMap<ItemStack, ItemStack> infusingList = new HashMap<ItemStack, ItemStack>();
  public HashMap<ItemStack, Integer> ticksList = new HashMap<ItemStack, Integer>();

  @SuppressWarnings("UnusedDeclaration")
  public void addRecipe(ItemStack input, ItemStack output, int tickTime) {
    infusingList.put(input, output);
    ticksList.put(input, tickTime);
  }

  public ItemStack getInfuserResult(ItemStack input) {
    for (Map.Entry<ItemStack, ItemStack> entry : infusingList.entrySet()) {
      if (ItemStack.areItemStacksEqual(entry.getKey(), input) && ItemStack.areItemStackTagsEqual(entry.getKey(), input)) return entry.getValue();
    }
    return null;
  }

  public int getInfuserTicks(ItemStack input) {
    for (Map.Entry<ItemStack, Integer> entry : ticksList.entrySet()) {
      if (ItemStack.areItemStacksEqual(entry.getKey(), input) && ItemStack.areItemStackTagsEqual(entry.getKey(), input)) return entry.getValue();
    }
    return 0;
  }
}
